turbo-txpool/src/lib.rs

```diff
@@ -91,114 +144,289 @@ impl Inner {
 
         ImportResult::Success
     }
+}
 
-    fn remove(&mut self, key: usize) {
-        let tx = self.txs.remove(key);
-        self.by_hash.remove(tx.hash()).expect("desync by hash");
+impl<C> Inner<C>
+where
+    C: Clone + Control,
+{
+    fn recheck(&mut self, diffs: &[AccountDiff]) {
+        #[derive(Default)]
+        struct Account {
+            balance: U256,
+            nonce: U256,
+        }
 
-        let removed = self.by_price.remove(&Priced {
-            gas_price: *tx.gas_price(),
-            key,
-        });
-        assert!(removed, "desync by price");
+        let mut accounts = HashMap::with_capacity(diffs.len());
+        for diff in diffs {
+            let account;
+            let address;
+
+            match &diff.diff {
+                Some(account_diff::Diff::Deleted(addr)) => {
+                    account = Account::default();
+                    address = Address::from_slice(&addr);
+                }
+                Some(account_diff::Diff::Changed(delta)) => {
+                    address = Address::from_slice(&delta.address);
+                    account = Account {
+                        balance: U256::from_big_endian(&delta.balance),
+                        nonce: U256::from_big_endian(&delta.nonce),
+                    };
+                }
+                None => continue,
+            };
+
+            accounts.insert(address, account);
+        }
+
+        let mut to_remove = Vec::new();
+        for (key, tx) in self.txs.iter() {
+            let account = match accounts.get(tx.from()) {
+                Some(a) => a,
+                None => continue,
+            };
+
+            if Self::validate(tx, account.nonce, account.balance).is_err() {
+                to_remove.push(key);
+            }
+        }
+
+        for key in to_remove.into_iter() {
+            self.remove(key);
+        }
     }
 
-    pub fn with_config(config: Config) -> Self {
-        Self {
-            max_txs: config.max_txs,
-            txs: Slab::with_capacity(config.max_txs),
-            by_hash: HashMap::with_capacity(config.max_txs),
-            by_price: BTreeSet::new(),
+    async fn block_reverted(&mut self, block: RevertedBlock) {
+        let reverted_hash = H256::from_slice(&block.reverted_hash);
+        let new_hash = H256::from_slice(&block.new_hash);
+
+        let replace = self
+            .latest_block
+            .map(|b| b == reverted_hash)
+            .unwrap_or(true);
+
+        if replace {
```

Comment:
This can probably be removed, since you can expect that to always receive the latest info that the provider is aware of.